### DISCUSSION TITLE
Derived predicates

### DISCUSSION THREAD

**Discussion by IgnacioVellido** [aiplan4eu/unified-planning#247] [Ideas]
Hi, I'm working on adding support for HPDL language (used by SIADEX-planner) and I would love if you could offer some guidance on how to add support for derived predicates.

Knowing that derived are syntactic sugar, could they be added as an extension of ```model.Fluent``` or a new class would be preferred?

An example of usage could be:

```python
from unified_planning.shortcuts import *

train = UserType('train')

c1 = Fluent('train-has-guard', BoolType(), t=train)
c2 = Fluent('train-has-driver', BoolType(), t=train)

d = Derived(name="train-usable", params={"t": train})
d.add_fluent(c1) # Add both fluents to derived
d.add_fluent(c2)

a = InstantaneousAction("a")
a.add_precondition(d)
```

with the PDDL equivalent:

```pddl
(:derived (train-usable ?t - train)
    (and
        (train-has-guard ?t)
        (train-has-driver ?t)
    )
)
```

**Comment by mikand** [comment]
Quick question out of ignorance: do derived predicates add expressivity at all?

What is the difference with just creating a python variable/function as follows?
```python
from unified_planning.shortcuts import *

train = UserType('train')

c1 = Fluent('train-has-guard', BoolType(), t=train)
c2 = Fluent('train-has-driver', BoolType(), t=train)

def d(x):
    return And(c1(x), c2(x))

a = InstantaneousAction("a", t=train)
a.add_precondition(d(t))
```

**Reply by IgnacioVellido** [reply]
Indeed, no additional expressivity at all, it's just a way to reuse complex expressions in PDDL.

From the point of view of a UP user there is no clear advantage in defining a derived, as one can keep the expression in variables as you showed. That would be more oriented to someone transitioning from PDDL, where reusing a composite expression multiple times (or one with forall, when, exists...) with a derived gives you some readability.

So, in a case when we are parsing an already defined domain, do you think it will be best just to keep track of the derived on ```pddl_reader.py``` and substitute for the expression whenever needed?

**Reply by roeger** [reply]
I would advise against using derived predicates in this context, where they are really only syntactic sugar (i.e. can be replaced with an expression that does not use derived predicates).  This could have a negative impact on some planning systems that do not treat this simple scenario as a special case.

But adding derived predicates (for arbitrary stratified axiom sets) could still be interesting because they add expressive power to the formalism. The disadvantage is that we cannot trivially compile them away. I looked into the "In Defense of PDDL Axioms" paper by Thiébaux et al.: If the translated task may only grow polynomially, then there is no compilation where plan length increase is also only polynomial (unless EXPTIME=PSPACE). This is also true for simple DATALOG axioms. Only if the axioms are non-recursive, such a compilation is possible (in this case, we can allow DATALOG with negation and the plans get preserved). This is basically a minor extension of the situation from Ignacio's example: Derived predicates in formulas get replaced with fixed expressions (without derived predicates). If we give up  on the requirement that the plan length may not explode, we can compile the axioms away (but axiom evaluation will basically be done by a long sequence of action applications that corresponds to this evaluation).

So adding derived predicates to the UP would be valuable, but in my opinion only if there is sufficient support by the underlying engines. For Fast Downward, there are no great admissible heuristics (for optimal planning) with axiom support. Many inadmissible heuristics work with tasks with axioms but the documentation uses the wonderful phrase "axioms: supported (in the sense that the planner won't complain -- handling of axioms might be very stupid and even render the heuristic unsafe)". If an engine cannot natively handle axioms, it would probably be better to model the problem in another way, so including derived predicates in the UP could lead modelers in an undesirable direction.
